internal/pkg/gateway/api.go

```diff
@@ -427,11 +428,81 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
 		return nil, status.Errorf(codes.Unavailable, "no orderer nodes available")
 	}
 
+	logger := logger.With("txID", request.TransactionId)
+	config := gs.getChannelConfig(request.ChannelId)
+	if config.ChannelConfig().Capabilities().ConsensusTypeBFT() {
+		return gs.submitBFT(ctx, orderers, txn, config, logger)
+	} else {
+		return gs.submitNonBFT(ctx, orderers, txn, logger)
+	}
+}
+
+func (gs *Server) submitBFT(ctx context.Context, orderers []*orderer, txn *common.Envelope, config channelconfig.Resources, logger *flogging.FabricLogger) (*gp.SubmitResponse, error) {
+	// For BFT, we send transaction to ALL orderers
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	waitCh := make(chan *gp.ErrorDetail, len(orderers))
+	for _, o := range orderers {
+		go func(ord *orderer) {
+			logger.Infow("Sending transaction to orderer", "endpoint", ord.logAddress)
+			ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
+			defer cancel()
+			response, err := gs.broadcast(ctx, ord, txn)
+
+			if err != nil {
+				logger.Warnw("Error sending transaction to orderer", "endpoint", ord.logAddress, "err", err)
+				waitCh <- errorDetail(ord.endpointConfig, err.Error())
+			} else if status := response.GetStatus(); status == common.Status_SUCCESS {
+				logger.Infow("Successful response from orderer", "endpoint", ord.logAddress)
+				waitCh <- nil
+			} else {
+				logger.Warnw("Unsuccessful response sending transaction to orderer", "endpoint", ord.logAddress, "status", status, "info", response.GetInfo())
+				waitCh <- errorDetail(ord.endpointConfig, fmt.Sprintf("received unsuccessful response from orderer: %s", common.Status_name[int32(status)]))
+			}
+		}(o)
+	}
+	n := len(orderers)
+	ordererConfig, ok := config.OrdererConfig()
+	if ok {
+		n = len(ordererConfig.Consenters())
+	}
+	quorum, _ := computeBFTQuorum(uint64(n))
+	successes, failures := 0, 0
+	var errDetails []proto.Message
+loop:
+	for i, total := 0, len(orderers); i < total; i++ {
+		select {
+		case osnErr := <-waitCh:
+			// Broadcast completed normally
+			if osnErr != nil {
+				errDetails = append(errDetails, osnErr)
+				failures++
+				if failures > total-quorum {
+					break loop
```

Comment:
I am not sure about "giving up" sending if we succeeded sending to less than a quorum.

I think we should still try to send to as many possible even if we failed to send to more than `f`. 

Probably it's best to do:
```
diff --git a/internal/pkg/gateway/api.go b/internal/pkg/gateway/api.go
index d2eebcabe..874c3a411 100644
--- a/internal/pkg/gateway/api.go
+++ b/internal/pkg/gateway/api.go
@@ -12,6 +12,8 @@ import (
        "io"
        "math/rand"
        "strings"
+       "sync/atomic"
+       "time"
 
        "github.com/golang/protobuf/proto"
        "github.com/hyperledger/fabric-protos-go/common"
@@ -438,6 +440,18 @@ func (gs *Server) Submit(ctx context.Context, request *gp.SubmitRequest) (*gp.Su
 }
 
 func (gs *Server) submitBFT(ctx context.Context, orderers []*orderer, txn *common.Envelope, config channelconfig.Resources, logger *flogging.FabricLogger) (*gp.SubmitResponse, error) {
+       var numFinishedSend uint32
+
+       earliestFinishDeadline := time.Now().Add(time.Millisecond * 500) // TODO: Probably should be configurable
+
+       n := len(orderers)
+
+       defer func() {
+               if atomic.LoadUint32(&numFinishedSend) < uint32(n) {
+                       time.Sleep(time.Until(earliestFinishDeadline))
+               }
+       }()
+
        // For BFT, we send transaction to ALL orderers
        ctx, cancel := context.WithCancel(ctx)
        defer cancel()
@@ -448,6 +462,7 @@ func (gs *Server) submitBFT(ctx context.Context, orderers []*orderer, txn *commo
                        ctx, cancel := context.WithTimeout(ctx, gs.options.BroadcastTimeout)
                        defer cancel()
                        response, err := gs.broadcast(ctx, ord, txn)
+                       atomic.AddUint32(&numFinishedSend, 1)
 
                        if err != nil {
                                logger.Warnw("Error sending transaction to orderer", "endpoint", ord.logAddress, "err", err)
@@ -461,7 +476,6 @@ func (gs *Server) submitBFT(ctx context.Context, orderers []*orderer, txn *commo
                        }
                }(o)
        }
-       n := len(orderers)
        ordererConfig, ok := config.OrdererConfig()
        if ok {
                n = len(ordererConfig.Consenters())

```

Reply:
I didn't intent to give up sending the transaction to all orderers, but rather to give up waiting for responses from orderers once consensus has been achieved.  But I can see how the code I had written could have that effect.  I've re-coded that bit.